[TableGen][DecoderEmitter] Rework table construction/emission

s-barannikov opened the pull request:
### Current state

We have FilterChooser class, which can be thought of as a **tree of encodings**. Tree nodes are instances of FilterChooser itself, and come in two types:

* A node containing single encoding that has *constant* bits in the specified bit range, a.k.a. singleton node.
* A node containing only child nodes, where each child represents a set of encodings that have the same *constant* bits in the specified bit range.

    Either of these nodes can have an additional child, which represents a set of encodings that have some *unknown* bits in the same bit range.

As can be seen, the **data structure is very high level**.

The encoding tree represented by FilterChooser is then converted into a finite-state machine (FSM), represented as **byte array**. The translation is straightforward: for each node of the tree we emit a sequence of opcodes that check encoding bits and predicates for each encoding. For a singleton node we also emit a terminal "decode" opcode.

The translation is done in one go, and this has negative consequences:

* We miss optimization opportunities.
* We have to use "fixups" when encoding transitions in the FSM since we don't know the size of the data we want to jump over in advance. We have to emit the data first and then fix up the location of the jump. This means the fixup size has to be large enough to encode the longest jump, so **most of the transitions are encoded inefficiently**.
* Finally, when converting the FSM into human readable form, we have to **decode the byte array we've just emitted**. This is also done in one go, so we **can't do any pretty printing**.

### This PR

We introduce an intermediary data structure, decoder tree, that can be thought as **AST of the decoder program**.
This data structure is **low level** and as such allows for optimization and analysis.
It resolves all the issues listed above. We now can:
* Emit more optimal opcode sequences.
* Compute the size of the data to be emitted in advance, avoiding fixups.
* Do pretty printing.

Serialization is done by a new class, DecoderTableEmitter, which converts the AST into a FSM in **textual form**, streamed right into the output file.

### Results
* The new approach immediately resulted in 12% total table size savings across all in-tree targets, without implementing any optimizations on the AST. Many tables observe ~20% size reduction.
* The generated file is much more readable.
* The implementation is arguably simpler and more straightforward (the diff is only +150~200 lines, which feels rather small for the benefits the change gives).

s-barannikov committed:
tmp

s-barannikov committed:
update tests

s-barannikov commented:
This is more or less ready for review.
(I'll need to write some description and add some comments.)

jurahul commented:
It's a bigger change, so @topperc would appreciate if you can review as well.

s-barannikov commented on llvm/utils/TableGen/DecoderEmitter.cpp:
```diff
@@ -1753,6 +1266,624 @@ static DecodeStatus decodeInstruction(const uint8_t DecodeTable[], MCInst &MI,
+      : DecoderTreeNode(CheckField), StartBit(StartBit), NumBits(NumBits),
+        Value(Value) {}
+
+  unsigned getStartBit() const { return StartBit; }
```
I have no clear rule for this in my personal unwritten coding standard. I group some closely-related members (as you can see down below), but not always.
These two are not very closely related, so my gut says they should stay visually separated.

s-barannikov commented on llvm/utils/TableGen/DecoderEmitter.cpp:
```diff
@@ -952,7 +641,8 @@ void FilterChooser::dumpStack(raw_ostream &OS, indent Indent,
 // decoded bits in order to verify that the instruction matches the Opcode.
 std::vector<FilterChooser::Island>
-FilterChooser::getIslands(const KnownBits &EncodingBits) const {
+FilterChooser::getIslands(const KnownBits &EncodingBits,
```
I'll extract this into separate PR. Not that it would make the diff much smaller, but this change is separable.

lenary commented:
What's the size/perf saving here? I presume it's mostly size, rather than perf?

s-barannikov commented:
> What's the size/perf saving here? I presume it's mostly size, rather than perf?

Currently, 12.1% reduction in size total, but more can be achieved in the future, since we now have a data structure we can do optimizations on. Performance should improve proportionally, but it can be more than 12% because the introduced Switch opcode works faster than a series of ExtractField + FilterValue.

<details>
<summary>raw data</summary>

```
| Target    | Old    | New    | Abs Diff | Rel Diff |
| --------- | ------ | ------ | -------- | -------- |
| AArch64   | 106857 |  92427 |   -14430 |  -13.50% |
| AArch64   |     63 |     58 |       -5 |   -7.94% |
| AMDGPU    |     12 |     12 |        0 |    0.00% |
| AMDGPU    |   6052 |   5029 |    -1023 |  -16.90% |
| AMDGPU    |  50472 |  45472 |    -5000 |   -9.91% |
| AMDGPU    |     24 |     22 |       -2 |   -8.33% |
| AMDGPU    |     69 |     64 |       -5 |   -7.25% |
| AMDGPU    |   6158 |   5099 |    -1059 |  -17.20% |
| AMDGPU    |  35890 |  31889 |    -4001 |  -11.15% |
| AMDGPU    |  14303 |  13212 |    -1091 |   -7.63% |
| AMDGPU    |   1528 |   1261 |     -267 |  -17.47% |
| AMDGPU    |  11164 |  10130 |    -1034 |   -9.26% |
| AMDGPU    |   5768 |   5330 |     -438 |   -7.59% |
| AMDGPU    |    124 |    114 |      -10 |   -8.06% |
| AMDGPU    |   6057 |   5021 |    -1036 |  -17.10% |
| AMDGPU    |    193 |    187 |       -6 |   -3.11% |
| AMDGPU    |    307 |    260 |      -47 |  -15.31% |
| AMDGPU    |   4938 |   4345 |     -593 |  -12.01% |
| AMDGPU    |  17481 |  16109 |    -1372 |   -7.85% |
| AMDGPU    |    227 |    189 |      -38 |  -16.74% |
| AMDGPU    |   3136 |   2881 |     -255 |   -8.13% |
| AMDGPU    |   1043 |    972 |      -71 |   -6.81% |
| AMDGPU    |  19855 |  17932 |    -1923 |   -9.69% |
| AMDGPU    |  43483 |  39886 |    -3597 |   -8.27% |
| AMDGPU    |   1524 |   1257 |     -267 |  -17.52% |
| AMDGPU    |  11663 |  10587 |    -1076 |   -9.23% |
| AMDGPU    |   5896 |   5440 |     -456 |   -7.73% |
| AMDGPU    |    608 |    566 |      -42 |   -6.91% |
| AMDGPU    |     78 |     74 |       -4 |   -5.13% |
| AMDGPU    |     24 |     24 |        0 |    0.00% |
| AMDGPU    |   5672 |   4716 |     -956 |  -16.85% |
| AMDGPU    |  20077 |  17904 |    -2173 |  -10.82% |
| AMDGPU    |     99 |     87 |      -12 |  -12.12% |
| AMDGPU    |   2324 |   2119 |     -205 |   -8.82% |
| AMDGPU    |   1886 |   1707 |     -179 |   -9.49% |
| AMDGPU    |   6546 |   5523 |    -1023 |  -15.63% |
| AMDGPU    |  54229 |  49095 |    -5134 |   -9.47% |
| AMDGPU    |     13 |     13 |        0 |    0.00% |
| AMDGPU    |   9903 |   8714 |    -1189 |  -12.01% |
| AMDGPU    |    247 |    209 |      -38 |  -15.38% |
| AMDGPU    |     83 |     77 |       -6 |   -7.23% |
| AMDGPU    |   1218 |   1060 |     -158 |  -12.97% |
| AMDGPU    |     13 |     13 |        0 |    0.00% |
| AMDGPU    |     49 |     47 |       -2 |   -4.08% |
| AMDGPU    |  13586 |  12305 |    -1281 |   -9.43% |
| AMDGPU    |     44 |     40 |       -4 |   -9.09% |
| ARC       |   1056 |    832 |     -224 |  -21.21% |
| ARC       |   2359 |   1871 |     -488 |  -20.69% |
| ARC       |     80 |     70 |      -10 |  -12.50% |
| ARC       |   1109 |    968 |     -141 |  -12.71% |
| ARM       |   5886 |   5252 |     -634 |  -10.77% |
| ARM       |    476 |    423 |      -53 |  -11.13% |
| ARM       |  16946 |  15451 |    -1495 |   -8.82% |
| ARM       |  13621 |  11845 |    -1776 |  -13.04% |
| ARM       |    305 |    279 |      -26 |   -8.52% |
| ARM       |   4934 |   4440 |     -494 |  -10.01% |
| ARM       |    809 |    699 |     -110 |  -13.60% |
| ARM       |     11 |     11 |        0 |    0.00% |
| ARM       |   6346 |   5751 |     -595 |   -9.38% |
| ARM       |    621 |    565 |      -56 |   -9.02% |
| ARM       |    509 |    427 |      -82 |  -16.11% |
| ARM       |     43 |     41 |       -2 |   -4.65% |
| ARM       |    241 |    197 |      -44 |  -18.26% |
| ARM       |    360 |    335 |      -25 |   -6.94% |
| ARM       |   1486 |   1366 |     -120 |   -8.08% |
| ARM       |   3902 |   3466 |     -436 |  -11.17% |
| ARM       |   2161 |   1960 |     -201 |   -9.30% |
| AVR       |   1036 |    865 |     -171 |  -16.51% |
| AVR       |     62 |     56 |       -6 |   -9.68% |
| AVR       |     21 |     19 |       -2 |   -9.52% |
| BPF       |   1583 |   1273 |     -310 |  -19.58% |
| BPF       |    222 |    185 |      -37 |  -16.67% |
| CSKY      |    792 |    652 |     -140 |  -17.68% |
| CSKY      |   4014 |   3460 |     -554 |  -13.80% |
| CSKY      |   2128 |   1866 |     -262 |  -12.31% |
| Hexagon   |  22424 |  19309 |    -3115 |  -13.89% |
| Hexagon   |   9675 |   8383 |    -1292 |  -13.35% |
| Hexagon   |   1900 |   1687 |     -213 |  -11.21% |
| Hexagon   |    363 |    317 |      -46 |  -12.67% |
| Hexagon   |     27 |     25 |       -2 |   -7.41% |
| Hexagon   |    218 |    188 |      -30 |  -13.76% |
| Hexagon   |     27 |     25 |       -2 |   -7.41% |
| Hexagon   |    123 |    109 |      -14 |  -11.38% |
| Lanai     |    910 |    743 |     -167 |  -18.35% |
| LoongArch |  20037 |  16007 |    -4030 |  -20.11% |
| M68k      |  11721 |  10004 |    -1717 |  -14.65% |
| Mips      |    426 |    352 |      -74 |  -17.37% |
| Mips      |    346 |    305 |      -41 |  -11.85% |
| Mips      |    349 |    298 |      -51 |  -14.61% |
| Mips      |     29 |     27 |       -2 |   -6.90% |
| Mips      |     41 |     35 |       -6 |  -14.63% |
| Mips      |    450 |    375 |      -75 |  -16.67% |
| Mips      |   4648 |   3849 |     -799 |  -17.19% |
| Mips      |     23 |     21 |       -2 |   -8.70% |
| Mips      |    243 |    208 |      -35 |  -14.40% |
| Mips      |    306 |    257 |      -49 |  -16.01% |
| Mips      |   2321 |   1960 |     -361 |  -15.55% |
| Mips      |     73 |     65 |       -8 |  -10.96% |
| Mips      |  12152 |  10270 |    -1882 |  -15.49% |
| Mips      |     37 |     33 |       -4 |  -10.81% |
| Mips      |   1764 |   1496 |     -268 |  -15.19% |
| Mips      |    151 |    135 |      -16 |  -10.60% |
| Mips      |     23 |     21 |       -2 |   -8.70% |
| Mips      |    183 |    159 |      -24 |  -13.11% |
| Mips      |     31 |     29 |       -2 |   -6.45% |
| Mips      |    910 |    778 |     -132 |  -14.51% |
| Mips      |     23 |     21 |       -2 |   -8.70% |
| Mips      |   1186 |   1037 |     -149 |  -12.56% |
| MSP430    |    271 |    222 |      -49 |  -18.08% |
| MSP430    |     90 |     74 |      -16 |  -17.78% |
| MSP430    |    264 |    217 |      -47 |  -17.80% |
| MSP430    |    253 |    207 |      -46 |  -18.18% |
| MSP430    |    253 |    207 |      -46 |  -18.18% |
| MSP430    |    253 |    207 |      -46 |  -18.18% |
| MSP430    |    471 |    375 |      -96 |  -20.38% |
| MSP430    |    429 |    339 |      -90 |  -20.98% |
| MSP430    |    563 |    467 |      -96 |  -17.05% |
| MSP430    |    541 |    447 |      -94 |  -17.38% |
| PowerPC   |  18930 |  16003 |    -2927 |  -15.46% |
| PowerPC   |   2112 |   1920 |     -192 |   -9.09% |
| PowerPC   |   3841 |   3347 |     -494 |  -12.86% |
| RISCV     |    717 |    611 |     -106 |  -14.78% |
| RISCV     |  20300 |  16937 |    -3363 |  -16.57% |
| RISCV     |     66 |     58 |       -8 |  -12.12% |
| RISCV     |    931 |    798 |     -133 |  -14.29% |
| RISCV     |    599 |    522 |      -77 |  -12.85% |
| RISCV     |   4294 |   3680 |     -614 |  -14.30% |
| RISCV     |    168 |    152 |      -16 |   -9.52% |
| RISCV     |    115 |    103 |      -12 |  -10.43% |
| RISCV     |    268 |    234 |      -34 |  -12.69% |
| RISCV     |   1391 |   1203 |     -188 |  -13.52% |
| RISCV     |    416 |    368 |      -48 |  -11.54% |
| RISCV     |    184 |    168 |      -16 |   -8.70% |
| RISCV     |     16 |     16 |        0 |    0.00% |
| RISCV     |     34 |     32 |       -2 |   -5.88% |
| RISCV     |   1224 |   1102 |     -122 |   -9.97% |
| RISCV     |    330 |    300 |      -30 |   -9.09% |
| RISCV     |   1806 |   1597 |     -209 |  -11.57% |
| RISCV     |     40 |     38 |       -2 |   -5.00% |
| RISCV     |    129 |    115 |      -14 |  -10.85% |
| RISCV     |    183 |    162 |      -21 |  -11.48% |
| RISCV     |     36 |     34 |       -2 |   -5.56% |
| RISCV     |    356 |    310 |      -46 |  -12.92% |
| RISCV     |   1152 |    986 |     -166 |  -14.41% |
| RISCV     |     26 |     24 |       -2 |   -7.69% |
| Sparc     |   6249 |   5290 |     -959 |  -15.35% |
| Sparc     |    166 |    148 |      -18 |  -10.84% |
| Sparc     |    322 |    288 |      -34 |  -10.56% |
| SystemZ   |    698 |    532 |     -166 |  -23.78% |
| SystemZ   |   8535 |   7104 |    -1431 |  -16.77% |
| SystemZ   |  18751 |  16002 |    -2749 |  -14.66% |
| VE        |  36002 |  32221 |    -3781 |  -10.50% |
| XCore     |   1300 |   1069 |     -231 |  -17.77% |
| XCore     |    923 |    793 |     -130 |  -14.08% |
| Xtensa    |    158 |    135 |      -23 |  -14.56% |
| Xtensa    |   3602 |   3074 |     -528 |  -14.66% |
|           | 777003 | 682869 |   -94134 |  -12.12% |

```

</details>

jurahul commented:
For the switch op, I was thinking we can also encode a sorted table of values to be searched directly in the table and then do a binary search on that table instead of a linear search today (still better than sequential ops). That means the values and sizes need to be encoded directly and not as ULEB128 (to avoid decoding cost)...

s-barannikov commented:
> That means the values and sizes need to be encoded directly and not as ULEB128 (to avoid decoding cost)...

The encoded sizes will occupy more space as well, since the distance to each next case will be the sum of the distance to previous case + its size (currently it is just size). But the idea is interesting.

jurahul commented on llvm/utils/TableGen/DecoderEmitter.cpp:
```diff
@@ -1718,6 +1245,619 @@ static DecodeStatus decodeInstruction(const uint8_t DecodeTable[], MCInst &MI,
+
+  // Calculate the number of decimal places for table indices.
+  // This is simply log10 of the table size.
+  IndexWidth = 1;
```
nit: llvm/include/llvm/DebugInfo/PDB/Native/FormatUtil.h has 

```
/// Returns the number of digits in the given integer.
inline int NumDigits(uint64_t N) {
```

maybe we can extract it into MathExtras.h or Format.h at a future point.

jurahul commented on llvm/utils/TableGen/DecoderEmitter.cpp:
```diff
@@ -1718,6 +1245,619 @@ static DecodeStatus decodeInstruction(const uint8_t DecodeTable[], MCInst &MI,
+                                               NegativeMask.getZExtValue()));
+  }
+
+  std::string DecoderIndex = getDecoderString(Encoding);
```
nit: var name should be DecoderString

jurahul commented on llvm/utils/TableGen/DecoderEmitter.cpp:
```diff
@@ -1718,6 +1245,619 @@ static DecodeStatus decodeInstruction(const uint8_t DecodeTable[], MCInst &MI,
+    unsigned Size = 0;
+    for (const DecoderTreeNode *Child : drop_end(N->children())) {
+      unsigned ChildSize = computeNodeSize(Child);
+      Size += 1 + getULEB128Size(ChildSize) + ChildSize;
```
can we add comments here to describe the additional count?

s-barannikov replied:
I added named constants, please see if it looks better.

s-barannikov committed:
Address review comments

* correct variable name
* remove redundant braces
* document integer literals in computeNodeSize
* add a common base class for CheckAllNode and CheckAnyNode

s-barannikov commented on llvm/utils/TableGen/DecoderEmitter.cpp:
```diff
@@ -1705,6 +1230,605 @@ static DecodeStatus decodeInstruction(const uint8_t DecodeTable[], MCInst &MI,
+  KindTy Kind;
+};
+
+class CheckManyNode : public DecoderTreeNode {
```
I couldn't think of a better name, suggestions are welcome

jurahul replied:
This seems ok. Resolving.

s-barannikov commented on llvm/utils/TableGen/DecoderEmitter.cpp:
```diff
@@ -1718,6 +1245,619 @@ static DecodeStatus decodeInstruction(const uint8_t DecodeTable[], MCInst &MI,
+DecoderTreeBuilder::buildCheckAnyNode(const FilterChooser &FC) {
+  auto N = std::make_unique<CheckAnyNode>();
+  if (FC.SingletonEncodingID) {
+    N->addChild(buildTerminalNode(*FC.SingletonEncodingID, FC.FilterBits));
```
Removed

s-barannikov commented on llvm/utils/TableGen/DecoderEmitter.cpp:
```diff
@@ -1718,6 +1245,619 @@ static DecodeStatus decodeInstruction(const uint8_t DecodeTable[], MCInst &MI,
+    N->addChild(buildCheckAllOrSwitchNode(FC.StartBit, FC.NumBits,
+                                          FC.FilterChooserMap));
+  }
+  if (FC.VariableFC) {
```
Removed

s-barannikov committed:
Rename DecoderTreeBuilder methods

The former names were hints for myself, give the methods more
appropriate names now.

s-barannikov committed:
Clarify size computation of CheckAny node

s-barannikov committed:
Avoid recursion when converting FilterChooser

Recursively converting FilterChooser results in "chained" CheckAny
nodes: the last child of each CheckAny node is another CheckAny node,
and each of them has no more than two children.

Avoid recursion and create single CheckAny node with multiple children.

There is no difference in output because the last child of CheckAny
node is printed unindented.

s-barannikov committed:
Print the last child of CheckNode indented

It should be indented like other children, because otherwise the output
looks as if it was a sibling, and those have different semantics.

Add a workaround for single-child CheckAny node not currently being
optimized. This compensates extra nesting caused by the above change.

s-barannikov committed:
Add a TODO with potential CheckAny optimization

s-barannikov commented on llvm/utils/TableGen/DecoderEmitter.cpp:
```diff
@@ -1705,6 +1230,626 @@ static DecodeStatus decodeInstruction(const uint8_t DecodeTable[], MCInst &MI,
+      N->addChild(convertFilterChooserMap(FC->StartBit, FC->NumBits,
+                                          FC->FilterChooserMap));
+    FC = FC->VariableFC.get();
+  } while (FC);
```
Changed recursion to a loop here. This should make it clear what VariableFC really is.

s-barannikov committed:
Revert "Add a TODO with potential CheckAny optimization"

This reverts commit 2d49938c2a8be4ed10f2392d01cc0665c24a776a.

Implemented suggestion results in increased table sizes on average.
The number of elements in scopes is too small for the suggestion to be
beneficial.

s-barannikov committed:
Extract DecoderTreeNode/DecoderTableEmitter into separate h/cpp files

Hopefully, logical separation simplifies the review.
I was going to do this in the next patch anyway.

s-barannikov committed:
Change `auto` to `const_iterator` to suppress clang-tidy warnings

s-barannikov committed:
Move/update comments, minor refactoring

s-barannikov commented on llvm/utils/TableGen/DecoderTableEmitter.cpp:
```diff
@@ -0,0 +1,342 @@
+  }
+  case DecoderTreeNode::CheckField: {
+    const auto *N = static_cast<const CheckFieldNode *>(Node);
+    return OpcodeSize + getULEB128Size(N->getStartBit()) + FieldWidthSize +
```
I think it is obvious enough. It would've been more obvious if size computation were combined with emitting. Unfortunately we need to know the size in advance.

jurahul commented on llvm/utils/TableGen/DecoderTableEmitter.cpp:
```diff
@@ -0,0 +1,342 @@
+}
+
+void DecoderTableEmitter::emitStartLine() {
+  CommentIndex = CurrentIndex;
```
CommentIndex seems to be really index of the start of this line. Its only used for printing in comment, but the comment itself is not a part of the table. Can we rename this to LineStartIndex to be more clear?

jurahul commented on llvm/utils/TableGen/DecoderTableEmitter.cpp:
```diff
@@ -0,0 +1,342 @@
+void DecoderTableEmitter::emitDecodeNode(const DecodeNode *N, indent Indent) {
+  const InstructionEncoding &Encoding = N->getEncoding();
+  unsigned InstOpcode = Encoding.getInstruction()->EnumVal;
+  unsigned DecoderIndex = TableInfo.getDecoderIndex(N->getDecoderString());
```
Q: Say we have several instructions that have the same decoder. It seems we would build the entire tree and duplicate the decoder string in each of these insts decode nodes, and only unique them at emission time. Is that correct? It may be possible to unique first at tree construction time itself so that there is no duplication (since DecoderString is stored as a std::string in the tree node). Not suggesting we do it now, but something for future if need be.

s-barannikov replied:
Good question.
I thought about uniquing them at tree construction time, but rejected the idea for some (probably puristic) reason.
I'll try to move it to the builder and see if there are any obstacles, thanks.

s-barannikov replied:
Yeah, I think it makes sense. At some point we might want to represent predicates/decoders in a more structured form, and that should be done at the construction stage.

s-barannikov replied:
It works, and simplifies the code quite a bit since we no longer need analyzeNode.
I'll try to add some comments and push after that.

jurahul replied:
Thanks. Resolving

s-barannikov commented on llvm/utils/TableGen/DecoderTableEmitter.cpp:
```diff
@@ -0,0 +1,342 @@
+
+void DecoderTableEmitter::emitCheckAllNode(const CheckAllNode *N,
+                                           indent Indent) {
+  for (const DecoderTreeNode *Child : N->children())
```
Yes, they are the same. FWIW there shouldn't be single-child CheckAll/CheckAny at all.
It may be useful to think of these nodes as C++ `||` and `&&` fold expressions.

I'll add a comment here with a TODO.

s-barannikov committed:
Print CheckAny as a sequence of try-else, fix SoftFail output

This reduces the vertical size of the output and should make it clearer.
Drop "check" when printing SoftFail because this opcode cannot fail.

s-barannikov committed:
Print single-bit bit range in more compact form

s-barannikov committed:
Rename CommentIndex -> LineStartIndex, comment index-related members

s-barannikov committed:
Add a TODO comment to emitCheckAllNode

s-barannikov committed:
Explain why the last child of CheckAny/SwitchField is emitted differently

s-barannikov commented on llvm/utils/TableGen/DecoderTableEmitter.cpp:
```diff
@@ -0,0 +1,342 @@
+      Size += getULEB128Size(Val) + getULEB128Size(ChildSize) + ChildSize;
+    }
+
+    auto [Val, Child] = *std::prev(N->case_end());
```
Added a comment here and to emitCheckAnyNode.

s-barannikov committed:
Move predicate/decoder unification to tree construction stage

s-barannikov committed:
Really add a comment

s-barannikov committed:
Remove InstructionEncoding from DecodeNode

We only need encoding name and instruction opcode.

jurahul commented:
Overall LGTM. @topperc or maybe @lenary if one of you can review as well, that would be great.

s-barannikov commented:
@jurahul Filled in the description, PTAL.

s-barannikov committed:
Document tree nodes

s-barannikov committed:
Merge branch 'main' into tablegen/decoder/tree

lenary reviewed (approved):
LGTM

topperc reviewed (approved):
LGTM

s-barannikov committed:
Merge branch 'main' into tablegen/decoder/tree
